### PR TITLE
feat: add TCPRoute controller

### LINKED ISSUE #16: TCPRoute Support (control-plane)
## Problem Statement

After we enable TCP support in the data-plane we should expose that support via the Gateway API resource [TCPRoute]().

## Acceptance Criteria

- [x] a `TCPRoute` controller is added which can compile `BACKEND` resources and push them to the `BACKENDS` map

### PR DESCRIPTION
**Pull request by shaneutt**
This primarily adds the `TCPRoute` controller and samples. Note however that the dataplane does not have complete TCP support yet (though some scaffolding for it is added as of this PR) so the dataplane configuration produced is valid but the samples wont work yet. This scaffolding however will help to support #45 and #15 which are responsible for completing this.

Resolves #16

This also includes some refactoring and cleanup, notably it refactors the dataplane code into modules.

### CODE REVIEW

File: controllers/tcproute_controller.go
Comment (mlavacca) [inline controllers/tcproute_controller.go:125]: I think we should check if the error is `NotFound`; in that case, keeping iterating to find a suitable gateway is fine. If we encounter another error, maybe we should immediately return it as an unexpected error occurred. WDYT?

File: controllers/tcproute_controller.go
Comment (mlavacca) [inline controllers/tcproute_controller.go:160]:
```suggestion
	referredGateway := &supportedGateways[0]
```

File: controllers/tcproute_controller.go
Comment (mlavacca) [inline controllers/tcproute_controller.go:71]:
```suggestion
	isManaged, gateway, err := r.isTCPRouteManaged(ctx, *tcproute)
	if err != nil {
		return ctrl.Result{}, err
	}
	if !isManaged {
		// TODO: enable orphan checking https://github.com/Kong/blixt/issues/47
		return ctrl.Result{}, nil
	}
```

Comment (mlavacca) [reply controllers/tcproute_controller.go:71]: For the sake of clarity, I think we should explicitly check the error here, because, as I get from the code, isManaged can be false with err == nil (no suitable gateway has been found) or with error != nil (an error occurred when getting the gateways, and no suitable gateway has been found).

File: controllers/tcproute_controller.go
Comment (mlavacca) [inline controllers/tcproute_controller.go:163]: Why do we want to return only the first Gateway of the list? Shouldn't we take into account all the matching Gateways?

File: controllers/tcproute_controller.go
Comment (mlavacca) [inline controllers/tcproute_controller.go:221]: WDYT about moving the finalizer removal out of this function? The function is named `ensureTCPRouteDeletedInDataPlane`, and removing also the dataplane finalizers looks a bit out of context.

File: tools/udp-test-server/Dockerfile
Comment (mlavacca) [inline tools/udp-test-server/Dockerfile:21]: Why are we moving this one from distroless to alpine?

**Review by mlavacca — COMMENTED** [review]
I left some comments on the controlplane part of the PR, I'll let @astoycos chime in and review the rust/eBPF side of things.

File: controllers/tcproute_controller.go
Comment (shaneutt) [reply controllers/tcproute_controller.go:221]: I think it's entirely relevant, the entire purpose of the finalizer is to hold up the object until we've confirmed removal from the dataplane.

File: controllers/tcproute_controller.go
Comment (shaneutt) [reply controllers/tcproute_controller.go]: (keep in mind that a lot of this controller code is copy/paste from the `UDPRoute` controller)

Yep this is a good suggestion, but it will enable a race condition that can cause a stale route so I've fixed that by adding a `Gateway`->`*Route` map func to both the `TCPRoute` and `UDPRoute` controllers to help compensate:

b1be1b75316230e4088b52fea6e5fe44082e5aaf
a6308da1f5e7e0de4cb95d6c6d6a569a76a8740d

And I've refactored both this method, and the method it was copy/pasted from in the `UDPRoute` controller:

f00e818f61908df9b9fa28d875ab29aa6a121c2f

And they no longer bother collecting errors anymore, any error that is not a 404 is considered a total failure to reconcile.

File: controllers/tcproute_controller.go
Comment (shaneutt) [reply controllers/tcproute_controller.go]: Another copy/paste from the old controller, now updated in both:

ffe1583221d679908b9ae7707bd7da7bbd6501db

File: controllers/tcproute_controller.go
Comment (shaneutt) [reply controllers/tcproute_controller.go]: In the code's current form, this wont make any technical difference, but I'm with you that the clarity could be better for the future. As this was copy/paste from the udproute controller, updated both in:

fa3af62270d21f1d6186ef4cac58e0b8ec13f273

File: controllers/tcproute_controller.go
Comment (shaneutt) [reply controllers/tcproute_controller.go]: Yes, but we're pacing ourselves to add support for multiple `Gateways` later as we don't need it right this moment for testing, and it provides an opportunity for a community member to jump in and work on something that's not huge. @rushi47 created https://github.com/Kong/blixt/issues/40 as a follow up for this, but it looks like we forgot to link that here. I've added the TODO with the link to both controllers in:

497953abbe09cd0052cf78086422b15092235bcc

File: controllers/tcproute_controller.go
Comment (shaneutt) [reply controllers/tcproute_controller.go:226]: I think it's entirely relevant, the entire purpose of the finalizer is to hold up the object until we've confirmed removal from the dataplane. Keep in mind too that now that we're moving this project to Kubernetes SIGs it is intended to become a reference implementation for Gateway API, and as such it's going to undergo a LOT of refactoring and reviews from Gateway API community members before we start advertising it as a reference implementation (in fact, the intention right now is that before we consider it "ready" as a reference implementation we will do a full PR review of the code with the org).

File: tools/udp-test-server/Dockerfile
Comment (shaneutt) [reply tools/udp-test-server/Dockerfile:21]: The commit for this included an explanation:

d783a56dc72de73b5b04b7099f3a09df175855be

```
This enables the ability to add tools, such as tcpdump
and do some debugging.
```

**Comment by shaneutt** [comment]
added `do-not-merge` as I intend to do some commit squashing in rebase before we merge.

**Review by mlavacca — APPROVED** [review]

**Review by mlavacca — DISMISSED** [review]

**Review by mlavacca — APPROVED** [review]

File: dataplane/ebpf/src/main.rs
Comment (astoycos) [inline dataplane/ebpf/src/main.rs:64]: I think it's still smarter to handle any errors we may encounter in our `handle` functions here rather than internally

Comment (astoycos) [reply dataplane/ebpf/src/main.rs:64]: (i.e like we still do for `tc_egress`)

Comment (astoycos) [reply dataplane/ebpf/src/main.rs:64]: Oh nm we do :) in `tc_ingress` my bad

File: Makefile
Comment (astoycos) [inline Makefile:275]:
```suggestion
	kubectl -n blixt-system rollout restart deployment blixt-controlplane
```

Comment (shaneutt) [reply Makefile:275]: This is a chained command with the previous line, thus why we're indenting as a visual cue

**Comment by shaneutt** [comment]
Re-based, ready to merge :+1:

**Pull request merged**
